src/main/java/io/appium/java_client/DefaultGenericMobileDriver.java

```diff
@@ -151,6 +157,32 @@ public List findElementsByXPath(String using) {
         return super.getMouse();
     }
 
+    /**
+     * @return a map with values that hold session details.
+     *
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public synchronized Map<String, Object> getSessionDetails() {
```

Comment:
I am against this change. Session details data is not immutable. It may change when driver is switched to another context (NATIVE APP or WEB VIEW). 

I think the right way is to find places where it is invoked frequently.

Reply:
If full copy is used, shouldn't there at least be an 'expiry time' or a method to force the details to be refreshed?

When I ran into this issue I decided to use an expiry time, based on the time it took to retrieve the information, in my (not so nice) workaround in https://github.com/fhoeben/hsac-fitnesse-appium.
The approach I currently use there is to have some cache at the element level, and using [my own 'json to element' converter](https://github.com/fhoeben/hsac-fitnesse-appium/blob/master/src/main/java/nl/hsac/fitnesse/fixture/util/mobile/element/MobileElementConverter.java). 
My custom elements require less retrieval of session details (for me the 'foundBy' information does not require session details, which is currently required for the driver's `toString()`, and they 'cache' information retrieved from the driver about an element for a short while (so my test code does not need to worry too much about for instance checking `isDisplayed()` twice)).